backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrolledSearch.java

```diff
@@ -19,94 +19,96 @@
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.search.SearchScrollRequest;
 import org.elasticsearch.client.RequestOptions;
-import org.elasticsearch.client.RestHighLevelClient;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.search.SearchHit;
 
 import com.github.fge.lambdas.Throwing;
 
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
 public class ScrolledSearch {
-    private static class ScrollIterator implements Iterator<SearchResponse>, Closeable {
-        private final RestHighLevelClient client;
-        private CompletableFuture<SearchResponse> searchResponseFuture;
 
-        ScrollIterator(RestHighLevelClient client, SearchRequest searchRequest) {
-            this.client = client;
-            ListenerToFuture<SearchResponse> listener = new ListenerToFuture<>();
-            client.searchAsync(searchRequest, RequestOptions.DEFAULT, listener);
+    private static final TimeValue TIMEOUT = TimeValue.timeValueMinutes(1);
 
-            this.searchResponseFuture = listener.getFuture();
-        }
+    private final ReactorElasticSearchClient client;
+    private final SearchRequest searchRequest;
 
-        @Override
-        public void close() throws IOException {
-            ClearScrollRequest clearScrollRequest = new ClearScrollRequest();
-            clearScrollRequest.addScrollId(searchResponseFuture.join().getScrollId());
-            client.clearScroll(clearScrollRequest, RequestOptions.DEFAULT);
-        }
+    public ScrolledSearch(ReactorElasticSearchClient client, SearchRequest searchRequest) {
+        this.client = client;
+        this.searchRequest = searchRequest;
+    }
 
-        @Override
-        public boolean hasNext() {
-            SearchResponse join = searchResponseFuture.join();
-            return !allSearchResponsesConsumed(join);
-        }
+    public Flux<SearchHit> searchHits() {
+        return searchResponses()
+            .concatMap(searchResponse -> Flux.just(searchResponse.getHits().getHits()));
+    }
 
-        @Override
-        public SearchResponse next() {
-            SearchResponse result = searchResponseFuture.join();
-            ListenerToFuture<SearchResponse> listener = new ListenerToFuture<>();
-            client.scrollAsync(
-                new SearchScrollRequest()
-                    .scrollId(result.getScrollId())
-                    .scroll(TIMEOUT),
-                RequestOptions.DEFAULT,
-                listener);
-            searchResponseFuture = listener.getFuture();
-            return result;
-        }
+    public Flux<SearchResponse> searchResponses() {
+        return Flux.push(sink -> {
+            AtomicReference<Optional<String>> scrollId = new AtomicReference<>(Optional.empty());
+            sink.onRequest(n -> {
+                if (n == 0) {
+                    sink.complete();
```

Comment:
it should only be `return` and not `complete`. We should also check of negative numbers. BTW, it should never happen.